SparseGrids/tsgCoreOneDimensional.hpp

```diff
@@ -205,19 +205,19 @@ namespace OneDimensionalMeta{
 namespace OneDimensionalNodes{
     // non-nested rules
     //! \brief Generate Gauss-Legendre weights \b w and points \b x for (input) number of points \b m.
-    void getGaussLegendre(int m, std::vector<double> &w, std::vector<double> &x);
+    void getGaussLegendre(int m, std::vector<double> &w, std::vector<double> &x, int version=1);
```

Comment:
Adding the variable like that is a bad way to test things. If you want to test different algorithms, you can make a pre-processor macro or a `constexpr` internal variable, then you make one change in the code, recompile and have it work both ways all throughout.

This is just more typing and less efficient. It doesn't matter in the end, since we will not merge this and will have only one decompose in the end.

Reply:
> Everything looks okay now, except for the Gauss-Laguerre tests, which are off by a small tolerance (~1e-6 for nodes and ~1e-12 for weights).

Gauss-Laguerre is one of the rules that works on an unbounded domain and it would make sense to be more sensitive. The question is whether your approach is more or less accurate than the original `decompose()`.

The original method has a tolerance parameter, can you check what happens if you make that one equal to machine `epsilon` or alternatively relax the tolerance in your approach to be `Maths::num_tol` which is about `1.E-12`.

Also, did you check Gauss-Chebyshev type 1? That weight is singular at the boundaries of the domain and I expect may be very sensitive to the eigenvalue method.

Reply:
> Adding the variable like that is a bad way to test things. If you want to test different algorithms, you can make a pre-processor macro or a `constexpr` internal variable, then you make one change in the code, recompile and have it work both ways all throughout.
> 
> This is just more typing and less efficient. It doesn't matter in the end, since we will not merge this and will have only one decompose in the end.

I'm not sure I understand. Wouldn't this force me to use only one of ``decompose()`` or ``decompose2()`` everywhere? Ideally, I would like to switch between the two based on a flag that I can change in the code, e.g., set the flag on and run ``getGaussLegendre()`` with ``decompose()``, then set the flag off and run ``getGaussLegendre()`` with ``decompose2()``, without recompiling in-between.

Reply:
> The original method has a tolerance parameter, can you check what happens if you make that one equal to machine `epsilon` or alternatively relax the tolerance in your approach to be `Maths::num_tol` which is about `1.E-12`.

Setting the original method's tolerance to machine `epsilon` and my method to `Maths::num_tol`, I saw a max error of `1.50067e-11` for the nodes and a max error of `5.77875e-09` for the weights at `n=1000` and `alpha=0.0`

> Also, did you check Gauss-Chebyshev type 1? That weight is singular at the boundaries of the domain and I expect may be very sensitive to the eigenvalue method.

The weights and nodes for all of the Gauss-Chebyshev variants are actually computed in closed form and don't need to invoke `decompose()`.

Reply:
> I'm not sure I understand. Wouldn't this force me to use only one of `decompose()` or `decompose2()` everywhere? Ideally, I would like to switch between the two based on a flag that I can change in the code, e.g., set the flag on and run `getGaussLegendre()` with `decompose()`, then set the flag off and run `getGaussLegendre()` with `decompose2()`, without recompiling in-between.

My idea was to recompile and keep the intrusion to the minimum. All will be fixed in the end, but don't want to forget something.

I also don't want default values in the internal API, I got bitten too many times. If a method is internal, e.g., used only by the developers, then each input should be specified. Doesn't matter for intermediate testing, but it does for the final code that makes into `master`.

Reply:
> The weights and nodes for all of the Gauss-Chebyshev variants are actually computed in closed form and don't need to invoke `decompose()`.

Gauss-Chebyshev type 1 is actually Gauss-Jacobi with `alpha = beta = -0.5`.

The closed form solution is perfect for testing the `decompose()` procedures, everything is best tested against a know closed form solution.